### ISSUE TITLE
Performance of CompDrop for other convnets

### ISSUE THREAD

**Issue by haruishi43** [LiheYoung/UniMatch-V2#7]
Excellent work!
I was curious to know if you have experimented with CompDrop or CompDrop+EMA on backbones such as ResNets and if the performance is superior to UniMatch. For example, how would UniMatchv2 perform with ResNet50 backbone on the Pascal VOC 1/16 (92) split?

**Comment by LiheYoung** [comment]
Hi, we did not conduct any experiments on the ResNet encoders, as one of our main motivations for this work is to switch from the outdated ResNet encoders to the most capable DINOv2 pre-trained encoders.

**Comment by haruishi43** [comment]
I think you did experiment on ResNet152 in table 9 right? Were there any other ConvNet based experiments?
And when you applied CompDrop to ResNet152, how does the dropout mask from binomial sampling work when the resolution of the features are different across the backbone layers?

**Comment by LiheYoung** [comment]
Oh, thank you for your reminder. Yes, the ResNet-152 result in Table 9 is obtained by our UniMatch V2 framework. We use these comparisons to demonstrate the superiority of DINOv2 encoders.

When applying CompDrop to ResNet-152, the dropping practice is indeed similar to DINOv2. The only difference is that we *independently* sample the dropping mask for each level of intermediate features. Below we provide a piece of code for you:
```python
dropout_masks = []
# the `features` is a list of intermediate features extracted from the encoder, it can be of different resolutions 
for feature in features:
    # the `feature` is the batch-level concatenated feature of images s1 and s2, something like: feature = encoder(torch.cat((s1, s2)))
    bs, dim = feature.shape[:2]

    # `dropout_mask1` is to drop features of image s1, and `dropout_mask2` is for features of image s2
    dropout_mask1 = self.binomial.sample((bs // 2, dim)).cuda() * 2.0
    dropout_mask2 = 2.0 - dropout_mask1
    # `dropout_prob` is to ensure only 50% training samples are performed CompDrop
    dropout_prob = 0.5
    num_kept = int(bs // 2 * (1 - dropout_prob))
    kept_indexes = torch.randperm(bs // 2)[:num_kept]
    dropout_mask1[kept_indexes, :] = 1.0
    dropout_mask2[kept_indexes, :] = 1.0
    
    dropout_mask = torch.cat((dropout_mask1, dropout_mask2))
            
    dropout_masks.append(dropout_mask)
          
features = (feature * dropout_masks[i][:, :, None, None] for i, feature in enumerate(features))
# then, we send the masked (i.e, dropped) `features` to the subsequent decoder
```

**Comment by haruishi43** [comment]
Thanks for the details!
Perhaps I might do some experiments with convnets to see the performance difference between CompDrop and Feature Perturbation+Strong View Perturbations.
